src/events/upgrade.py

```diff
@@ -120,15 +120,20 @@ def current_version(self) -> str:
         dependency_model: DependencyModel = getattr(self.dependency_model, "kafka_service")
         return dependency_model.version
 
+    @property
+    def zookeeper_current_version(self) -> str:
+        """Get current Zookeeper version."""
+        return self.charm.state.zookeeper.zookeeper_version
+
     @override
     def pre_upgrade_check(self) -> None:
-        if self.idle:
-            self._set_rolling_update_partition(partition=len(self.charm.state.brokers) - 1)
-
         default_message = "Pre-upgrade check failed and cannot safely upgrade"
         if not self.charm.healthy:
             raise ClusterNotReadyError(message=default_message, cause="Cluster is not healthy")
 
+        if self.idle:
+            self._set_rolling_update_partition(partition=len(self.charm.state.brokers) - 1)
```

Comment:
I believe this check and the patch of the K8s resource should be done after the healthy check. Otherwise we may raise an exception but apply the changes, which is somewhat counterintuitive

Reply:
Very good catch.